src/ext-test/java/org/opentripplanner/ext/stopconsolidation/StopConsolidationModuleTest.java

```diff
@@ -0,0 +1,45 @@
+package org.opentripplanner.ext.stopconsolidation;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.opentripplanner.ext.stopconsolidation.TestStopConsolidationModel.PATTERN;
+import static org.opentripplanner.ext.stopconsolidation.TestStopConsolidationModel.STOP_A;
+import static org.opentripplanner.ext.stopconsolidation.TestStopConsolidationModel.STOP_B;
+import static org.opentripplanner.ext.stopconsolidation.TestStopConsolidationModel.STOP_C;
+import static org.opentripplanner.ext.stopconsolidation.TestStopConsolidationModel.STOP_D;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import org.opentripplanner.ext.stopconsolidation.internal.DefaultStopConsolidationRepository;
+import org.opentripplanner.ext.stopconsolidation.model.ConsolidatedStopGroup;
+
+class StopConsolidationModuleTest {
+
+  @Test
+  void replacePatterns() {
+    var groups = List.of(new ConsolidatedStopGroup(STOP_D.getId(), List.of(STOP_B.getId())));
+
+    var transitModel = TestStopConsolidationModel.buildTransitModel();
+    transitModel.addTripPattern(PATTERN.getId(), PATTERN);
+    var repo = new DefaultStopConsolidationRepository();
+    var module = new StopConsolidationModule(transitModel, repo, groups);
+    module.buildGraph();
+
+    var modifiedPattern = transitModel.getTripPatternForId(PATTERN.getId());
+    assertFalse(modifiedPattern.getRoutingTripPattern().getPattern().sameAs(PATTERN));
+    assertFalse(modifiedPattern.sameAs(PATTERN));
+
+    var modifiedStop = modifiedPattern
+      .getRoutingTripPattern()
+      .getPattern()
+      .getStopPattern()
+      .getStop(1);
+    assertEquals(modifiedStop, STOP_D);
+
+    var patterns = List.copyOf(transitModel.getAllTripPatterns());
+    assertEquals(1, patterns.size());
```

Comment:
Consider: Doing test asserts on collection size is not recomended. You should test that the expected output is there, and if you want to test that something is not there, then you remove the expected and assert that the result is empty - printing the none empty result. This way, when the test fails you know why. Failing with index-out-of-bound is slightly better than this, because it is less code.

I frequently convert the entire collection to a string (sorting entities if necessary), this way I test everything above in just one single assert.

Reply:
(I wrote this comment in response to another comment, I have not read your code, and do not know what the test i about)

Reply:
I think that calling toString on a collection is a symptom of not having an expressive enough assert syntax. I think that this would be better solved through an assert library like assertj or truth which automatically gives you good failure messages.

Reply:
Adressed in f671f54e92444fa8533e5669f36ecd13ef00204e.